### PR TITLE
Code refactoring Aim and Speed

### PR DESCRIPTION
**Pull request by GoldenMine0502**
**this pr is for refactoring.**

**we could see simple aim and speed now.**

it seems to have slight difference on diff calc.

i could check two factors:
1. recent slider nerf
2. AimVelocity now starts to calculate second note, originally third.
.. etc? (if you find additional problem it is my problem)

overall i think its not fatal problem.

recent off by:
Test(1.448475413914554d,"zero-length-sliders")
    Expected: 1.4484754139145539d +/- 1.0000000000000001E-05d
    But was:  1.4804738333333043d
    Off by:   -0.031998419418750368d
    
TestClockRateAdjusted(1.7548875851757628d,"zero-length-sliders")
    Expected: 1.7548875851757628d +/- 1.0000000000000001E-05d
    But was:  1.806793929165774d
    Off by:   -0.05190634399001115d

TestClockRateAdjusted(8.938255920868981d,"diffcalc-test")
    Expected: 8.9382559208689809d +/- 1.0000000000000001E-05d
    But was:  8.9383248067074295d
    Off by:   -6.888583844855134E-05d

### CODE REVIEW

File: osu.Game/Rulesets/Difficulty/Skills/PreStrainSkill.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+using System.Linq;
+using System.Text;
+using osu.Game.Beatmaps;
+using osu.Game.Rulesets.Difficulty.Preprocessing;
+using osu.Game.Rulesets.Mods;
+
+namespace osu.Game.Rulesets.Difficulty.Skills
+{
+    public abstract class PreStrainSkill : Skill
+    {
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Difficulty/Skills/PreStrainSkill.cs:12]: We can't have a complete copy of all other skills in a new class hierarchy. Consider waiting on https://github.com/ppy/osu/pull/15010

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game/Rulesets/Difficulty/Skills/PreStrainSkill.cs (hunk 1 above)
Comment (GoldenMine0502) [reply osu.Game/Rulesets/Difficulty/Skills/PreStrainSkill.cs:12]: Because the preskills don't use necessarily in last calculation(OsuDifficultyCal/PerformanceCal) so the preskills may be confused. And we doesn't have pre calculation though we have current, post calculation. In this situation, If we want to use other skill result, we should put it before original calculation. This is why we should separate it to two types. in structure issue, this may be not good, but something like this requires to calculate note relations. As a result, the roles of two types are very different. it just uses same interface.

I can't find relations between this pr and the pr. Did he really write for calculating it?

Please be radical for changing overall structure. We should try all of radical things.

Comment (GoldenMine0502) [reply osu.Game/Rulesets/Difficulty/Skills/PreStrainSkill.cs:12]: If you mean just changing return types from PreStrainSkill[] to Skill[], it is reasonable.

Comment (smoogipoo) [reply osu.Game/Rulesets/Difficulty/Skills/PreStrainSkill.cs:12]: This can happen by passing `Skill`s into other `Skill`s instead. Your `PreStrainSkill` structure provides nothing extra or different to `StrainSkill`, it's a copy-paste of `StrainSkill` + `StrainDecaySkill`.

We already have too many classes. The path forward, as per #15010, is to separate out their defining features (sections, strain decay, etc) into classes which can be used to compose skills. The PR isn't perfect, but the theme is there.

Comment (GoldenMine0502) [reply osu.Game/Rulesets/Difficulty/Skills/PreStrainSkill.cs]: resolved

**Comment by GoldenMine0502** [comment]
i checked a profile to check whether this change affects on original pp result.

this is result refactor only for current live pp.

As a result, we could view just simple codes from Aim and Speed.

**Comment by GoldenMine0502** [comment]
at first i just want to put only PreStrainSkill.cs but i should provide an example as you said.

**Comment by GoldenMine0502** [comment]
it seems to have slight difference on diff calc.

i could check two factors:
1. recent slider nerf
2. AimVelocity now starts to calculate second note, originally third.
.. etc? (if you find additional problem it is my problem)

overall i think its not fatal problem.



recent off by:
Test(1.448475413914554d,"zero-length-sliders")
    Expected: 1.4484754139145539d +/- 1.0000000000000001E-05d
    But was:  1.4804738333333043d
    Off by:   -0.031998419418750368d
    
TestClockRateAdjusted(1.7548875851757628d,"zero-length-sliders")
    Expected: 1.7548875851757628d +/- 1.0000000000000001E-05d
    But was:  1.806793929165774d
    Off by:   -0.05190634399001115d
TestClockRateAdjusted(8.938255920868981d,"diffcalc-test")
    Expected: 8.9382559208689809d +/- 1.0000000000000001E-05d
    But was:  8.9383248067074295d
    Off by:   -6.888583844855134E-05d

File: osu.Game/Rulesets/Difficulty/Skills/PreStrainSkill.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+            CurrentSectionPeak += StrainValueAt(index, current);
+            SaveCurrentPeak();
+        }
+
+        protected override double CalculateInitialStrain(double time) => 0;
+
+        public void ProcessPre(int index, DifficultyHitObject current)
+        {
+            ProcessInternal(index, current);
+        }
```
Comment (bdach) [inline osu.Game/Rulesets/Difficulty/Skills/PreStrainSkill.cs:35]: i'm really confused by this. `ProcessPre()` calls `ProcessInternal()` which calls `Process()`. i do not understand why this exists when you may as well call `ProcessInternal()` directly?

also this index thing doesn't need to (and arguably shouldn't) exist. if skills need to be storing strain values from previous objects, you don't need to do any of this, just keep an internal list/limited size queue of the past values in the skill itself.

File: osu.Game/Rulesets/Difficulty/Skills/PreStrainSkill.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+using osu.Game.Rulesets.Difficulty.Preprocessing;
+using osu.Game.Rulesets.Mods;
+
+namespace osu.Game.Rulesets.Difficulty.Skills
+{
+    public abstract class PreStrainSkill : StrainSkill
+    {
+        protected abstract double StrainDecayBase { get; }
+
+        protected abstract double SkillMultiplier { get; }
```
Comment (bdach) [inline osu.Game/Rulesets/Difficulty/Skills/PreStrainSkill.cs:14]: nothing ever actually reads this value (same goes for `GetAllStrainPeaks()`)

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game/Rulesets/Difficulty/Skills/PreStrainSkill.cs (hunk 2 above)
Comment (GoldenMine0502) [reply osu.Game/Rulesets/Difficulty/Skills/PreStrainSkill.cs:35]: do you want to visibility ProcessInternal() from protected to public?

if then why do they set the visibility to internal? no reason to hide it now

Comment (bdach) [reply osu.Game/Rulesets/Difficulty/Skills/PreStrainSkill.cs:35]: if the only point of this added method is just to expose `ProcessInternal` as public, then yes, might as well

Comment (GoldenMine0502) [reply osu.Game/Rulesets/Difficulty/Skills/PreStrainSkill.cs:35]: yep

File: osu.Game/Rulesets/Difficulty/Skills/PreStrainSkill.cs (hunk 3 above)
Comment (GoldenMine0502) [reply osu.Game/Rulesets/Difficulty/Skills/PreStrainSkill.cs:14]: future feature but okay
after this the class role is not for strain

Comment (GoldenMine0502) [reply osu.Game/Rulesets/Difficulty/Skills/PreStrainSkill.cs:14]: for example using strain
https://github.com/GoldenMine0502/osu/blob/tapvariance/osu.Game.Rulesets.Osu/Difficulty/Skills/Pre/NoteVarianceAngle.cs

File: osu.Game/Rulesets/Difficulty/Skills/PreStrainSkill.cs (hunk 2 above)
Comment (GoldenMine0502) [reply osu.Game/Rulesets/Difficulty/Skills/PreStrainSkill.cs:35]: but we can't get difficultyvalue() if it doesnt store all values

Comment (GoldenMine0502) [reply osu.Game/Rulesets/Difficulty/Skills/PreStrainSkill.cs:35]: Ideal solution:
1. Keep the array storing all values
2. The overloading changes to StrainPeaks[StrainPeaks.Count - 1 - i]
2-1. Deleting all index parameter

I will do it until tomorrow.

**Comment by peppy** [comment]
Would appreciate if someone with more knowledge of this change can provide a better PR title. The current one is... hard to parse.

**Comment by bdach** [comment]
If I did have such knowledge I would have already changed it. The author of this pull is pretty much working solo, and due to a seemingly insurmountable communication barrier I'm having an extremely difficult time comprehending pretty much all of this thread.

**Comment by GoldenMine0502** [comment]
now this pr is not for the first conversation.

this change is just refactoring.

**Comment by GoldenMine0502** [comment]
A comment of the first conversation:

In the current pp system, we measure 'aim,' 'speed,' 'acc,' and 'flashlight' independently, and the problem arises here; the parts where 'aim' and 'speed' are mixed at a considerable level together, cannot be properly measured. (i.g. tech parts)

 We can think of A and B maps. A map has hard jumps in the front part and hard streams in the last part. The jumps and streams of the B map themselves are a little easier than the ones of the A map, but they appear simultaneously. So the overall difficulty we feel in the B map would be higher than the one in the A map. (these were what I wanted to express in the last image)

The point is, in the current system, the measured difficulty of the A map could be similar to what we feel, but the one of the B map would be much lower level to what we feel. We can't consider the synergistic effect that increases the actual difficulty when two ( or more ) elements are mixed now. I want to say this is because the 'aim' and 'speed' are measured independently. When relatively measured, we can alleviate this to some extent.

**Comment by smoogipoo** [comment]
- Why is there a "PreStrainSkill" class again? I thought I mentioned that I don't want a new class hierarchy here.
- `ProcessInternal` shouldn't be public. It's "internal" for a reason.
- This isn't really much of a refactoring, you've just added one more indirection but all the classes are still completely tied together. This doesn't for example, allow you to use the `AimVelocity` pre-strain-skill inside both the Speed and Aim skills.

I really don't see what we gain out of this... Am I alone in thinking this way?

**Comment by smoogipoo** [comment]
Sorry, but I'm going to take executive action here and close this PR as not the direction we want to go.

Feel free to start a [discussion](https://github.com/ppy/osu/discussions) and say your thoughts on where things should go / what you want to be able to do with the diffcalc system before taking on such refactorings.
